Mathlib/Probability/Kernel/Disintegration/Density.lean

```diff
@@ -0,0 +1,799 @@
+    {s s' : Set β} (h : s ⊆ s') :
+    densityProcess κ ν n a x s ≤ densityProcess κ ν n a x s' := by
+  unfold densityProcess
+  by_cases h0 : ν a (countablePartitionSet n x) = 0
+  · rw [h0, ENNReal.toReal_div, ENNReal.toReal_div]
+    simp
+  have h_ne_top : ∀ s,
+      κ a (countablePartitionSet n x ×ˢ s) / ν a (countablePartitionSet n x) ≠ ⊤ := by
+    intro s
+    rw [ne_eq, ENNReal.div_eq_top]
+    simp only [ne_eq, h0, and_false, false_or, not_and, not_not]
+    refine fun h_top ↦ eq_top_mono ?_ h_top
+    exact meas_countablePartitionSet_le_of_fst_le hκν n a x s
+  rw [ENNReal.toReal_le_toReal (h_ne_top s) (h_ne_top s')]
+  gcongr
+  simp [prod_subset_prod_iff, subset_rfl, h]
+
+lemma densityProcess_mono_kernel_left {κ' : kernel α (γ × β)} (hκκ' : κ ≤ κ')
+    (hκ'ν : fst κ' ≤ ν) (n : ℕ) (a : α) (x : γ) (s : Set β) :
+    densityProcess κ ν n a x s ≤ densityProcess κ' ν n a x s := by
+  unfold densityProcess
+  by_cases h0 : ν a (countablePartitionSet n x) = 0
+  · rw [h0, ENNReal.toReal_div, ENNReal.toReal_div]
+    simp
+  have h_le : κ' a (countablePartitionSet n x ×ˢ s) ≤ ν a (countablePartitionSet n x) :=
+    meas_countablePartitionSet_le_of_fst_le hκ'ν n a x s
+  rw [ENNReal.toReal_le_toReal]
+  · gcongr
+    exact hκκ' _ _
+  · rw [ne_eq, ENNReal.div_eq_top]
+    simp only [ne_eq, h0, and_false, false_or, not_and, not_not]
+    refine fun h_top ↦ eq_top_mono ?_ h_top
+    exact (hκκ' _ _).trans h_le
+  · rw [ne_eq, ENNReal.div_eq_top]
+    simp only [ne_eq, h0, and_false, false_or, not_and, not_not]
+    exact fun h_top ↦ eq_top_mono h_le h_top
+
+lemma densityProcess_antitone_kernel_right {ν' : kernel α γ}
+    (hνν' : ν ≤ ν') (hκν : fst κ ≤ ν) (n : ℕ) (a : α) (x : γ) (s : Set β) :
+    densityProcess κ ν' n a x s ≤ densityProcess κ ν n a x s := by
+  unfold densityProcess
+  have h_le : κ a (countablePartitionSet n x ×ˢ s) ≤ ν a (countablePartitionSet n x) :=
+    meas_countablePartitionSet_le_of_fst_le hκν n a x s
+  by_cases h0 : ν a (countablePartitionSet n x) = 0
+  · simp [le_antisymm (h_le.trans h0.le) zero_le', h0]
+  have h0' : ν' a (countablePartitionSet n x) ≠ 0 :=
+    fun h ↦ h0 (le_antisymm ((hνν' _ _).trans h.le) zero_le')
+  rw [ENNReal.toReal_le_toReal]
+  · gcongr
+    exact hνν' _ _
+  · simp only [ne_eq, ENNReal.div_eq_top, h0', and_false, false_or, not_and, not_not]
+    refine fun h_top ↦ eq_top_mono ?_ h_top
+    exact h_le.trans (hνν' _ _)
+  · simp only [ne_eq, ENNReal.div_eq_top, h0, and_false, false_or, not_and, not_not]
+    exact fun h_top ↦ eq_top_mono h_le h_top
+
+@[simp]
+lemma densityProcess_empty (κ : kernel α (γ × β)) (ν : kernel α γ) (n : ℕ) (a : α) (x : γ) :
+    densityProcess κ ν n a x ∅ = 0 := by
+  simp [densityProcess]
+
+lemma tendsto_densityProcess_atTop_empty_of_antitone (κ : kernel α (γ × β)) (ν : kernel α γ)
+    [IsFiniteKernel κ] (n : ℕ) (a : α) (x : γ)
+    (s : ℕ → Set β) (hs : Antitone s) (hs_iInter : ⋂ i, s i = ∅)
+    (hs_meas : ∀ n, MeasurableSet (s n)) :
```

Comment:
```suggestion
    (hs_meas : ∀ m, MeasurableSet (s m)) :
```
(keep `n` for the index of the countable partition). Same thing below several times.